doc/user/content/transform-data/troubleshooting.md

```diff
@@ -277,6 +277,310 @@ If you've gone through the dataflow troubleshooting and do not want to make any
+and other internal views to discover which parts of your query are
+computationally expensive (e.g.,
+[`mz_introspection.mz_compute_operator_durations_histogram`](/sql/system-catalog/mz_introspection/#mz_compute_operator_durations_histogram), [`mz_introspection.mz_scheduling_elapsed`](/sql/system-catalog/mz_introspection/#mz_scheduling_elapsed))
+or consuming excessive memory (e.g., ).
+
+We show how to use
+[`mz_introspection.mz_lir_mapping`](/sql/system-catalog/mz_introspection/#mz_lir_mapping)
+by example based on the [`Auction` load
+generator](/sql/create-source/load-generator/#auction) from the
+[quickstart guide](/get-started/quickstart/#step-2-create-the-source):
+
+```sql
+CREATE SOURCE auction_house
+FROM LOAD GENERATOR AUCTION
+(TICK INTERVAL '1s', AS OF 100000)
+FOR ALL TABLES;
+
+CREATE VIEW winning_bids AS
+  SELECT DISTINCT ON (a.id) b.*, a.item, a.seller
+    FROM auctions AS a
+    JOIN bids AS b
+      ON a.id = b.auction_id
+   WHERE b.bid_time < a.end_time
+     AND mz_now() >= a.end_time
+   ORDER BY a.id, b.amount DESC, b.bid_time, b.buyer;
+
+CREATE INDEX wins_by_item ON winning_bids (item);
+```
+
+We attribute four different kinds of performance data to parts of the
+`wins_by_item` query: computation time, memory usage, [top-k](/transform-data/idiomatic-materialize-sql/top-k/) query [sizing hints](/transform-data/optimization/#query-hints), and worker
+skew.
+
+### Attributing computation time
+
+One way to understand which parts of your query are 'expensive' is to
+look at the overall time spent computing in each part. Materialize
+reports the time spent in each _dataflow operator_ in
+[`mz_introspection.mz_compute_operator_durations_histogram`](/sql/system-catalog/mz_introspection/#mz_compute_operator_durations_histogram). By
+joining it with
+[`mz_introspection.mz_lir_mapping`](/sql/system-catalog/mz_introspection/#mz_lir_mapping),
+we can attribute the time spent in each operator to the higher-level,
+more intelligible LIR operators:
+
+```sql
+SELECT mo.name AS name, global_id, lir_id, parent_lir_id, REPEAT(' ', nesting * 2) || operator AS operator,
+       SUM(duration_ns)/1000 * '1 microsecond'::INTERVAL AS duration, SUM(count) AS count
+    FROM           mz_introspection.mz_lir_mapping mlm
+         LEFT JOIN mz_introspection.mz_compute_operator_durations_histogram mcodh
+                ON (mlm.operator_id_start <= mcodh.id AND mcodh.id < mlm.operator_id_end)
+              JOIN mz_catalog.mz_objects mo
+                ON (mlm.global_id = mo.id)
+GROUP BY mo.name, global_id, lir_id, operator, parent_lir_id, nesting
+ORDER BY global_id, lir_id DESC;
+```
+
+Running this query on an auction generator will produce results that look something like the following (though the specific numbers will vary, of course):
+
+
+| name         | global_id | lir_id | parent_lir_id | operator                         | duration        | count  |
```

Comment:
So, markdown table like this won't preserve the spacing(I know, with all your nice indentation logic ... come on, markdown)  :shakes-fist:  
https://preview.materialize.com/materialize/30899/transform-data/troubleshooting/#attributing-computation-time

When I get back, I can move separate these into a data file and a table 
where in the data file, can use ```mzsql annotation to maintain spacing.

Reply:
😭 Thanks!

I could also probably just force `&nbsp;` or a unicode non-breaking space in there or something, though that's a hell of a kludge. Awkward either way.